presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -686,6 +749,45 @@ private List<TaskSource> getTaskSources(Iterator<SerializedPrestoSparkTaskSource
         return result.build();
     }
 
+    private List<TaskSource> getNativeExecutionShuffleSources(
+            Session session, TaskId taskId, PlanFragment fragment, Map<PlanNodeId, PrestoSparkShuffleReadInfo> shuffleReadInfos, List<TaskSource> taskSources)
+    {
+        ImmutableSet.Builder<ScheduledSplit> result = ImmutableSet.builder();
+        PlanNode root = fragment.getRoot();
+        AtomicLong nextSplitId = new AtomicLong();
+        shuffleReadInfos.forEach((planNodeId, info) ->
+                result.add(new ScheduledSplit(nextSplitId.getAndIncrement(), planNodeId, new Split(REMOTE_CONNECTOR_ID, new RemoteTransactionHandle(), new RemoteSplit(
+                        new Location(format("batch://%s?shuffleInfo=%s", taskId, shuffleInfoTranslator.createSerializedReadInfo(info))),
+                        taskId)))));
+
+        List<TaskSource> nativeExecutionSources = taskSources.stream().filter(taskSource -> isNativeExecutionNode(taskSource.getPlanNodeId(), root)).collect(Collectors.toList());
+        checkArgument(nativeExecutionSources.size() <= 1, "At most 1 taskSource is expected for NativeExecutionNode but got %s", nativeExecutionSources.size());
+        if (!nativeExecutionSources.isEmpty()) {
+            // Append the shuffle splits with original splits
+            TaskSource nativeExecutionSource = nativeExecutionSources.get(0);
+            result.addAll(nativeExecutionSource.getSplits());
+        }
+
+        TaskSource newTaskSource = new TaskSource(root.getId(), result.build(), ImmutableSet.of(Lifespan.taskWide()), true);
+        ImmutableList.Builder<TaskSource> newTaskSources = ImmutableList.builder();
+        // Combine the shuffle read taskSource and original sources
+        newTaskSources.add(newTaskSource).addAll(taskSources.stream().filter(taskSource -> !isNativeExecutionNode(taskSource.getPlanNodeId(), root)).collect(Collectors.toList()));
+        return newTaskSources.build();
```

Comment:
This code is possibly incorrect.

There is immutableBuilder, some elements are added, then filtered and assigned to a new list, but the assigned list is discarded.

Then all the nodes are returned. 

If not incorrect, this code is very confusing.

Reply:
addAll() adds the filtered and collected result of taskSources (note that everything is in addAll() method call). Nothing is discarded here.